pkg/pprof/fix_go_heap_truncated.go

```diff
@@ -0,0 +1,266 @@
+package pprof
+
+import (
+	"bytes"
+	"reflect"
+	"sort"
+	"unsafe"
+
+	"golang.org/x/exp/slices"
+
+	profilev1 "github.com/grafana/pyroscope/api/gen/proto/go/google/v1"
+)
+
+const (
+	minGroupSize = 2
+
+	tokens    = 8
+	tokenLen  = 16
+	suffixLen = tokens + tokenLen
+
+	tokenBytesLen  = tokenLen * 8
+	suffixBytesLen = suffixLen * 8
+)
+
+// MayHaveGoHeapTruncatedStacktraces reports whether there are
+// any chances that the profile may have truncated stack traces.
+func MayHaveGoHeapTruncatedStacktraces(p *profilev1.Profile) bool {
+	if !hasGoHeapSampleTypes(p) {
+		return false
+	}
+	// Some truncated stacks have depth less than the depth limit (32).
+	const minDepth = 28
+	for _, s := range p.Sample {
+		if len(s.LocationId) >= minDepth {
+			return true
+		}
+	}
```

Comment:
See https://go-review.googlesource.com/c/go/+/572396: we also need to check that the depth does not exceed the limit